### PR TITLE
Conformance between spec and API: kmeans

### PR DESCRIPTION
**Pull request by mchernov-intel**

### CODE REVIEW

**Comment by mchernov-intel** [comment]
@amgrigoriev please, review

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/infer_kernel_dpc.cpp [hunk 1]
```diff
@@ -40,10 +40,11 @@ using daal_kmeans_lloyd_dense_ucapi_kernel_t =
...
 template <typename Float>
-struct infer_kernel_gpu<Float, method::by_default> {
-    infer_result operator()(const dal::backend::context_gpu& ctx,
-                            const descriptor_base& params,
-                            const infer_input& input) const {
+struct infer_kernel_gpu<Float, method::by_default, task::clustering> {
+    infer_result<task::clustering> operator()(const dal::backend::context_gpu& ctx,
+                                              const descriptor_base<task::clustering>& params,
+                                              const infer_input<task::clustering>& input) const {
+        using Task = task::clustering;
```
Comment (amgrigoriev) [inline cpp/oneapi/dal/algo/kmeans/backend/gpu/infer_kernel_dpc.cpp:47]: This alias is just for a single occurrence. Maybe it ins't necessary?

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/train_kernel_lloyd_dense_dpc.cpp [hunk 2]
```diff
@@ -41,10 +41,11 @@ using daal_kmeans_lloyd_dense_ucapi_kernel_t =
...
 template <typename Float>
-struct train_kernel_gpu<Float, method::lloyd_dense> {
-    train_result operator()(const dal::backend::context_gpu& ctx,
-                            const descriptor_base& params,
-                            const train_input& input) const {
+struct train_kernel_gpu<Float, method::lloyd_dense, task::clustering> {
+    train_result<task::clustering> operator()(const dal::backend::context_gpu& ctx,
+                                              const descriptor_base<task::clustering>& params,
+                                              const train_input<task::clustering>& input) const {
+        using Task = task::clustering;
```
Comment (amgrigoriev) [inline cpp/oneapi/dal/algo/kmeans/backend/gpu/train_kernel_lloyd_dense_dpc.cpp:48]: The same comment about the alias

File: cpp/oneapi/dal/algo/kmeans/common.cpp [hunk 3]
```diff
@@ -19,68 +19,84 @@
...
+class detail::model_impl<task::clustering> : public base {
 public:
     table centroids;
 };
 
 using detail::descriptor_impl;
 using detail::model_impl;
 
-descriptor_base::descriptor_base() : impl_(new descriptor_impl{}) {}
+template <typename Task>
```
Comment (amgrigoriev) [inline cpp/oneapi/dal/algo/kmeans/common.cpp:39]: Task -> task::clustering?

File: cpp/oneapi/dal/algo/kmeans/common.cpp [hunk 4]
```diff
@@ -19,68 +19,84 @@
...
+descriptor_base<Task>::descriptor_base() : impl_(new descriptor_impl{}) {}
 
-std::int64_t descriptor_base::get_cluster_count() const {
+template <typename Task>
+std::int64_t descriptor_base<Task>::get_cluster_count() const {
```
Comment (amgrigoriev) [inline cpp/oneapi/dal/algo/kmeans/common.cpp:43]: Task -> task::clustering?

File: cpp/oneapi/dal/algo/kmeans/common.cpp [hunk 5]
```diff
@@ -19,68 +19,84 @@
...
     return impl_->cluster_count;
 }
 
-std::int64_t descriptor_base::get_max_iteration_count() const {
+template <typename Task>
+std::int64_t descriptor_base<Task>::get_max_iteration_count() const {
     return impl_->max_iteration_count;
```
Comment (amgrigoriev) [inline cpp/oneapi/dal/algo/kmeans/common.cpp:49]: Task -> task::clustering?

File: cpp/oneapi/dal/algo/kmeans/common.cpp [hunk 6]
```diff
@@ -19,68 +19,84 @@
...
 }
 
-double descriptor_base::get_accuracy_threshold() const {
+template <typename Task>
+double descriptor_base<Task>::get_accuracy_threshold() const {
```
Comment (amgrigoriev) [inline cpp/oneapi/dal/algo/kmeans/common.cpp:53]: Task -> task::clustering?

File: cpp/oneapi/dal/algo/kmeans/common.cpp [hunk 7]
```diff
@@ -19,68 +19,84 @@
...
     return impl_->accuracy_threshold;
 }
 
-void descriptor_base::set_cluster_count_impl(std::int64_t value) {
+template <typename Task>
+void descriptor_base<Task>::set_cluster_count_impl(std::int64_t value) {
```
Comment (amgrigoriev) [inline cpp/oneapi/dal/algo/kmeans/common.cpp:58]: Task -> task::clustering?

File: cpp/oneapi/dal/algo/kmeans/common.cpp [hunk 8]
```diff
@@ -19,68 +19,84 @@
...
     if (value <= 0) {
         throw domain_error("cluster_count should be > 0");
     }
     impl_->cluster_count = value;
 }
 
-void descriptor_base::set_max_iteration_count_impl(std::int64_t value) {
+template <typename Task>
+void descriptor_base<Task>::set_max_iteration_count_impl(std::int64_t value) {
```
Comment (amgrigoriev) [inline cpp/oneapi/dal/algo/kmeans/common.cpp:66]: Task -> task::clustering?

File: cpp/oneapi/dal/algo/kmeans/common.cpp [hunk 9]
```diff
@@ -19,68 +19,84 @@
...
     if (value < 0) {
         throw domain_error("max_iteration_count should be >= 0");
     }
     impl_->max_iteration_count = value;
 }
 
-void descriptor_base::set_accuracy_threshold_impl(double value) {
+template <typename Task>
+void descriptor_base<Task>::set_accuracy_threshold_impl(double value) {
```
Comment (amgrigoriev) [inline cpp/oneapi/dal/algo/kmeans/common.cpp:74]: The same

File: cpp/oneapi/dal/algo/kmeans/common.cpp [hunk 10]
```diff
@@ -19,68 +19,84 @@
...
     if (value < 0.0) {
         throw domain_error("accuracy_threshold should be >= 0.0");
     }
     impl_->accuracy_threshold = value;
 }
 
-model::model() : impl_(new model_impl{}) {}
+template <typename Task>
+model<Task>::model() : impl_(new model_impl{}) {}
```
Comment (amgrigoriev) [inline cpp/oneapi/dal/algo/kmeans/common.cpp:82]: The same comment for model class

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/infer_kernel_dpc.cpp (hunk 1 above)
Comment (mchernov-intel) [reply cpp/oneapi/dal/algo/kmeans/backend/gpu/infer_kernel_dpc.cpp:47]: fixed

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/train_kernel_lloyd_dense_dpc.cpp (hunk 2 above)
Comment (mchernov-intel) [reply cpp/oneapi/dal/algo/kmeans/backend/gpu/train_kernel_lloyd_dense_dpc.cpp:48]: fixed

File: cpp/oneapi/dal/algo/kmeans/common.cpp (hunk 4 above)
Comment (mchernov-intel) [reply cpp/oneapi/dal/algo/kmeans/common.cpp]: fixed

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/infer_kernel_dpc.cpp (hunk 1 above)
Comment (amgrigoriev) [inline cpp/oneapi/dal/algo/kmeans/backend/gpu/infer_kernel_dpc.cpp:47]: Unused identifier Task?

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/train_kernel_lloyd_dense_dpc.cpp (hunk 2 above)
Comment (amgrigoriev) [inline cpp/oneapi/dal/algo/kmeans/backend/gpu/train_kernel_lloyd_dense_dpc.cpp:48]: Unused "Task"?

**Review by amgrigoriev — APPROVED** [review]

**Pull request merged**